File: web/js/lib/bulk-plan.js
/**
 * Bulk-download split semantics (WP 4a.3).
 *
 * Ports the mockup's `syncBulk` primary/secondary decision (docs/design/
 * vault-app-mockup-NOTES.md round 5, "Bulk actions never silently
 * re-download a cached game"): multi-select classifies the picked set by
 * REAL cache state and targets only what needs bytes — the button spells
 * out the skip count, re-download is always an explicit secondary, never
 * folded into the primary. Pure — no DOM, no fetch — the classification
 * and the resulting button/note text are both unit-testable
 * (web/tests/bulk-plan.test.js).
 *
 * Narrower than the mockup by exactly the same divergence documented in
 * game-status.js: there is no "stale" status, so the mockup's three-way
 * split (free / stale-update / cached) collapses to two buckets here —
 * `needsDownload` (not cached OR errored — see the docstring on
 * `classifyBulkSelection` for why `error` joins `none`) and `current`
 * (cached). `busy` (already has a job in flight) is excluded from both, as
 * in the mockup.
 *
 * `classifyBulkDeleteEligibility` (WP 4a.3 review fix, should-fix 1) lives
 * here too, next to the download split, rather than inlined in
 * `views/library.js` where the review round found it untested: the
 * mockup's own delete-eligibility rule is has-cache-content
 * (`vault-app-mockup.html`'s `dels = rest.filter(g=>g.st!=="none")`), which
 * for the real API means `hasVisibleCacheContent`, NOT "not none" —
 * `dispKind(g) !== "none"` also matches `error`, and an errored app with
 * ZERO bytes has no depot mappings left to delete
 * (`DELETE /v1/cache/{appid}` 404s: "appid has no depot_app_map rows").
 */

import { dispKind, findLiveJob, hasVisibleCacheContent, KIND } from "./game-status.js";

/** Appids with a prefill job that is queued, running or paused right now —
 * shared by both classifiers below (queued jobs count as busy here, unlike
 * `findLiveJob` — dedupe protection needs the queued case too; mockup
 * parity: `isBusy` checks `LIVE.includes(...) || st==="queued"`). */
function busyAppidsFromJobs(jobs) {
  return new Set(
    (Array.isArray(jobs) ? jobs : [])
      .filter((j) => j.type === "prefill" && ["queued", "running", "paused"].includes(j.status))
      .map((j) => j.appid),
  );
}

/**
 * @param {object[]} games the SELECTED games (already resolved from the
 *   picked appid set — callers own that lookup).
 * @param {object[] | null | undefined} jobs `GET /v1/jobs` snapshot.
 * @returns {{
 *   busy: object[], needsDownload: object[], current: object[],
 * }}
 */
export function classifyBulkSelection(games, jobs) {
  const busyAppids = busyAppidsFromJobs(jobs);
  const busy = [];
  const rest = [];
  for (const g of games) (busyAppids.has(g.appid) ? busy : rest).push(g);

  const needsDownload = [];
  const current = [];
  for (const g of rest) {
    // A live job can't be true here (busyAppids already excludes it), so
    // dispKind's cache-only branch is exactly what we want: `none` and
    // `error` both mean "not successfully cached" — an errored app gets
    // the same "needs a (re)download" treatment as a never-downloaded one,
    // matching statusAction's retry decision in game-status.js.
    const kind = dispKind(g, undefined);
    (kind === KIND.CACHED ? current : needsDownload).push(g);
  }
  return { busy, needsDownload, current };
}

/**
 * Which of the SELECTED games can actually be sent to
 * `DELETE /v1/cache/{appid}` without a guaranteed 404 — has real bytes on
 * the cache right now, AND is not busy (deleting under an in-flight prefill
 * is a 409 anyway; excluding it here keeps the button's own count honest
 * about what it will actually attempt). Deliberately
 * `hasVisibleCacheContent`, not `dispKind(...) !== "none"`: an `error`
 * status with zero visible bytes (e.g. a first-ever prefill that failed
 * before writing anything) has no depot mappings to delete — offering it
 * for bulk delete would just 404. An `error` status WITH bytes (a half-
 * deleted or partially-failed run, api/README.md "Per-game deletion":
 * "a failed depot is typically half deleted") genuinely has content to
 * clean up and stays eligible.
 *
 * @param {object[]} games the SELECTED games.
 * @param {object[] | null | undefined} jobs `GET /v1/jobs` snapshot.
 * @returns {object[]} the subset of `games` eligible for bulk delete.
 */
export function classifyBulkDeleteEligibility(games, jobs) {
  const busyAppids = busyAppidsFromJobs(jobs);
  return games.filter((g) => !busyAppids.has(g.appid) && hasVisibleCacheContent(g));
}

const plural = (n, noun) => `${n} ${noun}${n === 1 ? "" : "s"}`;

/**
 * Build the bulk-download bar's button/note text from a classification.
 * Mirrors the mockup's three visible outcomes (mockup's `stale` branch is
 * gone — see module header):
 *   1. Something needs downloading -> primary targets exactly that,
 *      skip count spelled out.
 *   2. Nothing needs downloading, nothing is `current` either (every pick
 *      is `busy`) -> primary disabled, "Already downloading".
 *   3. Nothing needs downloading, everything is `current` -> primary
 *      disabled ("All cached"), secondary offers an explicit re-download.
 *
 * @param {{busy: object[], needsDownload: object[], current: object[]}} classification
 * @param {number} totalPicked
 */
export function buildBulkDownloadPlan(classification, totalPicked) {
  const { busy, needsDownload, current } = classification;

  if (needsDownload.length) {
    const skipped = totalPicked - needsDownload.length;
    return {
      primaryEnabled: true,
      primaryLabel:
        needsDownload.length < totalPicked
          ? `Download ${needsDownload.length} of ${totalPicked}`
          : `Download ${plural(needsDownload.length, "game")}`,
      primaryTargets: needsDownload.map((g) => g.appid),
      note: skipped
        ? `${skipped} already cached — not re-downloaded.`
        : needsDownload.length > 1
          ? `All ${needsDownload.length} are new to the cache.`
          : "",
      secondaryLabel: null,
      secondaryTargets: [],
    };
  }

  if (current.length) {
    return {
      primaryEnabled: false,
      primaryLabel: "All cached — nothing to download",
      primaryTargets: [],
      note: "Every selected game is current. Re-download only if you need to refetch from Steam.",
      secondaryLabel: `Re-download ${current.length}`,
      secondaryTargets: current.map((g) => g.appid),
    };
  }

  return {
    primaryEnabled: false,
    primaryLabel: busy.length === totalPicked && totalPicked > 0 ? "Already downloading" : "Nothing to download",
    primaryTargets: [],
    note:
      busy.length === totalPicked && totalPicked > 0
        ? "Every selected game already has a job in flight."
        : "",
    secondaryLabel: null,
    secondaryTargets: [],
  };
}

// findLiveJob is re-exported for callers that already have this module open
// and want the same "live job" notion classifyBulkSelection is built on top
// of (game-status.js remains the canonical source).
export { findLiveJob };

File: app/app/src/main/java/dev/steamvault/app/ui/library/logic/BulkPlan.kt
package dev.steamvault.app.ui.library.logic

import dev.steamvault.app.net.model.GameSummary
import dev.steamvault.app.net.model.JobSummary
import dev.steamvault.app.ui.status.StatusKind

/**
 * Bulk-download split semantics (WP 4b.4) — Kotlin port of
 * `web/js/lib/bulk-plan.js`'s classify/plan pair (docs/design/
 * vault-app-mockup-NOTES.md round 5, "Bulk actions never silently
 * re-download a cached game"): multi-select classifies the picked set by
 * REAL cache state and targets only what needs bytes — the button spells
 * out the skip count, re-download is always an explicit secondary, never
 * folded into the primary. Pure — no Compose, no network — the
 * classification and the resulting label/note text are both unit-testable
 * (`BulkPlanTest`, mirroring `web/tests/bulk-plan.test.js`'s named cases).
 *
 * Narrower than the mockup by exactly the same divergence documented in
 * `GameStatus.kt`: no "stale" status, so the mockup's three-way split
 * (free / stale-update / cached) collapses to two buckets —
 * `needsDownload` (not cached OR errored — see [classifyBulkSelection]'s
 * kdoc for why `error` joins `none`) and `current` (cached). `busy`
 * (already has a job in flight) is excluded from both, as in the mockup.
 *
 * [classifyBulkDeleteEligibility] lives here too, next to the download
 * split, rather than inlined in the Library screen (same review-fix shape
 * the web port's WP 4a.3 review applied): the mockup's own delete-
 * eligibility rule is has-cache-content, which for the real API means
 * [hasVisibleCacheContent], NOT "kind != none" — `dispKind(g) != NONE` also
 * matches ERROR, and an errored app with ZERO bytes has no depot mappings
 * left to delete (`DELETE /v1/cache/{appid}` 404s: "appid has no
 * depot_app_map rows").
 *
 * **String-resource rule, exception recorded (WP 4b.4 review fix; see
 * app/README.md's "String resources" conventions section for the general
 * policy).** The button labels/notes built in [buildBulkDownloadPlan]
 * (`"Download %d of %d"`, `"All cached — nothing to download"`,
 * `"%d already cached — not re-downloaded."`, etc.) stay Kotlin string
 * literals here, NOT `strings.xml` resources, on purpose: they are a
 * line-for-line, word-for-word port of `web/js/lib/bulk-plan.js`'s own
 * literal strings (see that file's `buildBulkDownloadPlan`), and the value
 * of that port is exactly its ABILITY TO DIFF against the web source one
 * clause at a time — routing through resource indirection (a `strings.xml`
 * entry plus format-arg plumbing) would obscure that diff and make the next
 * person to touch either side re-derive the correspondence by hand instead
 * of reading it off two side-by-side literals. This is a narrow,
 * deliberate exception: it applies to strings whose entire reason for
 * existing is "the exact wording bulk-plan.js already decided", not to
 * general UI chrome (which does belong in resources — see `LibraryStrings.kt`
 * for the analogous dynamic-toast case and why THAT one could not simply
 * mirror this same exception).
 */

/** Appids with a prefill job that is queued, running or paused right now —
 * shared by both classifiers below (queued jobs count as busy here, unlike
 * [findLiveJob] — dedupe protection needs the queued case too; mockup
 * parity: a job counts as busy if it is queued, running or paused). */
private fun busyAppidsFromJobs(jobs: List<JobSummary>): Set<Int> =
    jobs.filter { it.type == "prefill" && it.status in setOf("queued", "running", "paused") }
        .map { it.appid }
        .toSet()

data class BulkSelectionClassification(
    val busy: List<GameSummary>,
    val needsDownload: List<GameSummary>,
    val current: List<GameSummary>,
)

/** @param games the SELECTED games (already resolved from the picked appid set — callers own that lookup). */
fun classifyBulkSelection(games: List<GameSummary>, jobs: List<JobSummary>): BulkSelectionClassification {
    val busyAppids = busyAppidsFromJobs(jobs)
    val busy = mutableListOf<GameSummary>()
    val rest = mutableListOf<GameSummary>()
    for (g in games) (if (g.appid in busyAppids) busy else rest).add(g)

    val needsDownload = mutableListOf<GameSummary>()
    val current = mutableListOf<GameSummary>()
    for (g in rest) {
        // A live job can't be true here (busyAppids already excludes it), so
        // dispKind's cache-only branch is exactly what we want: NONE and
        // ERROR both mean "not successfully cached" -- an errored app gets
        // the same "needs a (re)download" treatment as a never-downloaded
        // one, matching statusAction's retry decision in GameStatus.kt.
        val kind = dispKind(g, null)
        (if (kind == StatusKind.CACHED) current else needsDownload).add(g)
    }
    return BulkSelectionClassification(busy, needsDownload, current)
}

/**
 * Which of the SELECTED games can actually be sent to
 * `DELETE /v1/cache/{appid}` without a guaranteed 404 — has real bytes on
 * the cache right now, AND is not busy (deleting under an in-flight prefill
 * is a 409 anyway). Deliberately [hasVisibleCacheContent], not
 * `dispKind(...) != NONE`: an ERROR status with zero visible bytes has no
 * depot mappings to delete; an ERROR status WITH bytes (a half-deleted or
 * partially-failed run) genuinely has content to clean up and stays
 * eligible.
 * @param games the SELECTED games.
 */
fun classifyBulkDeleteEligibility(games: List<GameSummary>, jobs: List<JobSummary>): List<GameSummary> {
    val busyAppids = busyAppidsFromJobs(jobs)
    return games.filter { it.appid !in busyAppids && hasVisibleCacheContent(it) }
}

private fun plural(n: Int, noun: String): String = "$n $noun" + if (n == 1) "" else "s"

/** The bulk-download bar's three visible outcomes (mockup round 5, narrowed
 * by one branch since there is no "stale" bucket — see module kdoc). */
data class BulkDownloadPlan(
    val primaryEnabled: Boolean,
    val primaryLabel: String,
    val primaryTargets: List<Int>,
    val note: String,
    val secondaryLabel: String?,
    val secondaryTargets: List<Int>,
)

fun buildBulkDownloadPlan(classification: BulkSelectionClassification, totalPicked: Int): BulkDownloadPlan {
    val (busy, needsDownload, current) = classification

    if (needsDownload.isNotEmpty()) {
        val skipped = totalPicked - needsDownload.size
        return BulkDownloadPlan(
            primaryEnabled = true,
            primaryLabel = if (needsDownload.size < totalPicked) {
                "Download ${needsDownload.size} of $totalPicked"
            } else {
                "Download ${plural(needsDownload.size, "game")}"
            },
            primaryTargets = needsDownload.map { it.appid },
            note = when {
                skipped > 0 -> "$skipped already cached — not re-downloaded."
                needsDownload.size > 1 -> "All ${needsDownload.size} are new to the cache."
                else -> ""
            },
            secondaryLabel = null,
            secondaryTargets = emptyList(),
        )
    }

    if (current.isNotEmpty()) {
        return BulkDownloadPlan(
            primaryEnabled = false,
            primaryLabel = "All cached — nothing to download",
            primaryTargets = emptyList(),
            note = "Every selected game is current. Re-download only if you need to refetch from Steam.",
            secondaryLabel = "Re-download ${current.size}",
            secondaryTargets = current.map { it.appid },
        )
    }

    val allBusy = busy.size == totalPicked && totalPicked > 0
    return BulkDownloadPlan(
        primaryEnabled = false,
        primaryLabel = if (allBusy) "Already downloading" else "Nothing to download",
        primaryTargets = emptyList(),
        note = if (allBusy) "Every selected game already has a job in flight." else "",
        secondaryLabel = null,
        secondaryTargets = emptyList(),
    )
}
